platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/transform/TransformedVariantConverter.java

```diff
@@ -0,0 +1,85 @@
+import org.gradle.api.internal.artifacts.transform.Transform;
+import org.gradle.api.internal.artifacts.transform.TransformStep;
+import org.gradle.api.internal.artifacts.transform.TransformedVariant;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
+
+import javax.inject.Inject;
+import java.util.List;
+
+/**
+ * This type is responsible for converting from heavyweight {@link TransformedVariant} instances to
+ * lightweight {@link TransformationChainData} instances.
+ * <p>
+ * See the {@link org.gradle.internal.component.resolution.failure.transform package javadoc} for why.
+ */
+public final class TransformedVariantConverter {
+    private final ImmutableAttributesFactory attributesFactory;
+
+    @Inject
+    public TransformedVariantConverter(ImmutableAttributesFactory attributesFactory) {
+        this.attributesFactory = attributesFactory;
+    }
+
+    public ImmutableList<TransformationChainData> convert(List<TransformedVariant> transformedVariants) {
+        ImmutableList.Builder<TransformationChainData> builder = ImmutableList.builder();
+        transformedVariants.forEach(transformedVariant -> builder.add(convert(transformedVariant)));
+        return builder.build();
+    }
+
+    private TransformationChainData convert(TransformedVariant transformedVariant) {
+        AttributeChangeRecordingVisitor visitor = new AttributeChangeRecordingVisitor(transformedVariant.getRoot());
+        transformedVariant.getTransformChain().visitTransformSteps(visitor);
+        return visitor.buildResultData();
+    }
+
+    private final class AttributeChangeRecordingVisitor implements Action<TransformStep> {
+        private final ImmutableList.Builder<TransformData> stepsBuilder = ImmutableList.builder();
+
+        private final SourceVariantData sourceVariant;
+        private ImmutableAttributes currentAttributes;
+
+        public AttributeChangeRecordingVisitor(ResolvedVariant root) {
+            sourceVariant = new SourceVariantData(root.asDescribable().getDisplayName(), root.getAttributes());
+            currentAttributes = sourceVariant.getAttributes();
+        }
+
+        @Override
+        public void execute(TransformStep transformStep) {
+            TransformData transformData = convert(currentAttributes, transformStep);
+            stepsBuilder.add(transformData);
+            currentAttributes = transformData.getFromAttributes();
+        }
+
+        public TransformationChainData buildResultData() {
+            return new TransformationChainData(sourceVariant, stepsBuilder.build());
+        }
+
+        private TransformData convert(ImmutableAttributes currentAttributes, TransformStep step) {
+            Transform transform = step.getTransform();
+            ImmutableAttributes resultingToAttributes = attributesFactory.join(currentAttributes, transform.getToAttributes()).asImmutable();
```

Comment:
If we go with representing the `TransformStep` instances in the final failure data, then we can avoid all this concating altogether. 

The final attributes are already pre-concated and provided by `TransformedVariant#getAttributes`

Reply:
> Despite the mutability of either {@code fallback} or {@code primary}, this method will always return a non-immutable container. All mutable operations are forwarded to the primary. 

So, it promises to always returns a non-immutable (but explicitly avoids saying mutable) container, since:

> Therefore, if the primary container is immutable, mutable operations are not supported on this container.

So...we've discovered [a container that is neither mutable or immutable](https://en.wikipedia.org/wiki/G%C3%B6del%27s_incompleteness_theorems).  Cool.

Reply:
> If we go with representing the TransformStep instances in the final failure data, then we can avoid all this concating altogether.

I don't want to try that.  Those contain some heavier-weight instances that we shouldn't expose as part of the failure data, it makes sense to use a completely new and minimal type, despite the effort of converting to it.

Reply:
I'm not proposing representing the entire transform

Reply:
What we have here is each node in the chain represents a variant. The "To" attributes represent the _result_ of applying the transform step to the previous transformed variant. So, we concat each way. 

I am simply recommending each link in the chain represents a _transform step_, not the result of applying the transform to the previous variant. That way, we don't need to recompute the concat'd attributes.

I am not proposing we track any extra state on each step or anything like that. Just that we change the semantics of each step in the chain, so we don't need to compute the attributes here

Reply:
As for the immutable vs mutable, it may be a mutable container. 

BUT its guaranteed to be an `AttributeContainerInternal`, which we can call `toImmutable` on

Reply:
OK, I see what you're saying.  I didn't set it up that way because:

The final set of _resulting_ attributes (starting with the source variant's attributes and applying all transforms in order) is important.  If we only record the _changing_ attributes at each step, we'd need to either:

1. Compute the final attributes later - doing 2 iterations of the chain (once to record each step, and once to compute)
2. Or store both the final step's to attributes and the final resulting attributes separately, which seemed unnecessary

So even though the final attributes are already available, this just seemed a little nicer to me.

With the deduplication testing work to follow this, however, I'm going to need just the exact to attributes for each transform anyway, not the computed attributes of the "virtual variant" at each step.  I think the computed attributes make slightly more sense to provide to Failure Describers that may want to investigate this situation, but they can always just recompute the intermediate states by following the chain from the source if they want to.

So I'll make this change and just duplicate the final attributes.